ext/storage_sources/s3_store/aws_bucket_conn.cxx

```diff
@@ -0,0 +1,114 @@
+#include <aws/core/Aws.h>
+#include <aws/s3-crt/S3CrtClient.h>
+#include <aws/s3-crt/model/DeleteObjectRequest.h>
+#include <aws/s3-crt/model/ListObjectsRequest.h>
+#include <aws/s3-crt/model/Object.h>
+#include <aws/s3-crt/model/PutObjectRequest.h>
+#include "aws_bucket_conn.h"
+
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+/*
+ * list_buckets --
+ *     Builds a list of buckets from AWS account into a vector. Returns true if success, otherwise
+ *     false.
+ */
+bool
+aws_bucket_conn::list_buckets(std::vector<std::string> &buckets) const
+{
+    auto outcome = m_s3_crt_client.ListBuckets();
+    if (outcome.IsSuccess()) {
+        for (const auto &bucket : outcome.GetResult().GetBuckets())
+            buckets.push_back(bucket.GetName());
+        return true;
+    } else {
+        std::cerr << "Error in list_buckets: " << outcome.GetError().GetMessage() << std::endl
+                  << std::endl;
+        return false;
+    }
+}
+
+/*
+ * list_objects --
+ *     Builds a list of objects from a S3 bucket into a vector. Returns true if success, otherwise
+ *     false.
+ */
+bool
+aws_bucket_conn::list_objects(
+  const std::string &bucket_name, std::vector<Aws::S3Crt::Model::Object> &bucket_objects) const
+{
+    Aws::S3Crt::Model::ListObjectsRequest request;
+    request.WithBucket(bucket_name);
+    Aws::S3Crt::Model::ListObjectsOutcome outcomes = m_s3_crt_client.ListObjects(request);
+
+    if (outcomes.IsSuccess()) {
+        bucket_objects = outcomes.GetResult().GetContents();
+        return true;
+    } else {
+        std::cerr << "Error in list_buckets: " << outcomes.GetError().GetMessage() << std::endl;
+        return false;
+    }
+}
+
+/*
+ * put_object --
+ *     Puts an object into an S3 bucket. Returns true if success, otherwise false.
+ */
+bool
+aws_bucket_conn::put_object(
+  const std::string &bucket_name, const std::string &object_key, const std::string &file_name) const
+{
+    Aws::S3Crt::Model::PutObjectRequest request;
+    request.SetBucket(bucket_name);
+    request.SetKey(object_key);
+
+    std::shared_ptr<Aws::IOStream> input_data = Aws::MakeShared<Aws::FStream>(
+      "s3-source", file_name.c_str(), std::ios_base::in | std::ios_base::binary);
+
+    if (!input_data->good()) {
```

Comment:
Do we need this check, or would we fail at PutObject call and handle the error there ?

Reply:
This can be removed.